src/odemis/driver/piezomotor.py

```diff
@@ -0,0 +1,681 @@
+            if self._axis_modes[axis] == CLOSED_LOOP:
+                encoder_cnts = round(val / ENCODER_STEPSIZE)
+                speed_usteps = round(self._speed / self._wfm_stepsizes[axis] * USTEPS_PER_WFM)  # steps / second
+                self._sendCommand(b'X%dT%d,%d' % (self._axis_map[axis], encoder_cnts, speed_usteps))
+                targets[axis] = encoder_cnts
+            else:
+                self._updatePosition()
+                target = val - self.position.value[axis]
+                wfm_steps = int(target / self._wfm_stepsizes[axis])  # number of waveform steps
+                usteps = int((target % self._wfm_stepsizes[axis]) * USTEPS_PER_WFM)   # number of µsteps
+                self._sendCommand(b'X%dJ%d,%d,%d' % (self._axis_map[axis], wfm_steps, usteps, speed_usteps))
+        self._waitEndMotion(targets)
+        self._updatePosition()
+
+    def _doMoveRel(self, shift):
+        targets = {}
+        self._updatePosition()
+        for axis, val in shift.items():
+            speed_usteps = round(self._speed / self._wfm_stepsizes[axis] * USTEPS_PER_WFM)  # steps / second
+            if self._axis_modes[axis] == CLOSED_LOOP:
+                encoder_cnts = round(val / ENCODER_STEPSIZE)
+                # There are two possibilities: move relative to current position (XC) and move relative to
+                # target position (XR). We are moving relative to the current position (might be more intuitive
+                # if something goes wrong and we're stuck in the wrong position).
+                self._sendCommand(b'X%dC%d,%d' % (self._axis_map[axis], encoder_cnts, speed_usteps))
+                targets[axis] = self.position.value[axis] / ENCODER_STEPSIZE + encoder_cnts
+            else:
+                wfm_steps = int(val / self._wfm_stepsizes[axis])  # number of waveform steps
+                usteps = int((val % self._wfm_stepsizes[axis]) * USTEPS_PER_WFM)   # number of µsteps
+                self._sendCommand(b'X%dJ%d,%d,%d' % (self._axis_map[axis], wfm_steps, usteps, speed_usteps))
+                targets[axis] = None
+        self._waitEndMotion(targets)
+        self._updatePosition()
+
+    def _waitEndMotion(self, targets):
+        """
+        Wait until move is done
+        :arg target (int or None): target (for closed-loop), None for open loop
+        """
+        for ax_name, t in targets.items():
+            if self._axis_modes[ax_name] == CLOSED_LOOP:
+                if t is None:
+                    raise ValueError("No target provided for closed-loop move on axis %s." % ax_name)
+                query_cmd = b'X%dE' % self._axis_map[ax_name]
+                resp_finished = b'%d' % t  # wait till encoder position reports target position
+            else:
+                query_cmd = b'X%dJ' % self._axis_map[ax_name]
+                resp_finished = b'1'  # 1 if finished, 0 while moving
+
+            moving = True
+            end_time = time.time() + 5  # 5 s until timeout
```

Comment:
That timeout is really crude. In addition, 5 s could be really short for some long moves, with closed loop. If you don't know better use 30s.
See the tmcm driver. Based on the distance and speed of each axis, it computes a rough estimation of how long it takes for each axis, and use the longest time.